Documentation/Open Location Code API Specification.md

```diff
@@ -0,0 +1,78 @@
+| `encode`             | Latitude, longitude             | The Plus Code with code length 10 with area that contains the specified coordinates |
+| `decode`             | Full code Plus Code string with length 10 | The southern parallel and western meridian for the Plus Code area |
+
+Implementations are RECOMMENDED to use the method names above.
+
+Note: it is possible to implement `isFull` using a Perl Compatible Regular Expression, see the Plus Code specification.
+
+## OPTIONAL public methods
+
+An implementation MAY implement these methods.
+
+| Method name        | Input(s)                                                     | Output(s)                                                    |
+| ------------------ | ------------------------------------------------------------ | ------------------------------------------------------------ |
+| `isValid`          | Plus Code string                                             | True if this is a Plus Code (full code or short code), false otherwise |
+| `isShort`          | Plus Code string                                             | True if this is a short code Plus Code, false otherwise      |
+| `encode`           | Latitude, longitude, code length                             | The full code Plus Code with specified code length with area that contains the specified coordinates |
+| `decode`           | Full code Plus Code string                                   | The southern parallel and western meridian for the Plus Code area |
+| `shorten`          | Full code Plus Code, reference location latitude and longitude | The short code Plus Code representing the input where all significant digits that are allowed to be omitted are omitted; or if no digits are allowed to be omitted then the input is returned |
+| `recoverNearest`   | Short code Plus Code, reference location latitude and longitude | The unique full code Plus Code nearest (latitudinal distance + longitudinal distance) to the reference location |
+
+Implementations are RECOMMENDED to use the method names above. Note that the OPTIONAL `decode` public method overrides and tightens the specification for the REQUIRED `decode` public method. Note that The OPTIONAL `encode` public method adds a parameter. This MAY be implemented as an overloaded method or another way that is customary for the programming language.
+
+An implementation MAY implement `shorten` using separate `shortenBy4` and `shortenBy6` methods.
+
+Note: it is possible to implement `isValid` and `isShort` using a Perl Compatible Regular Expression, see the Plus Code specification.
+
+## REQUIREMENTS for all public methods
+
+Implementations MUST follow these REQUIREMENTS for all public methods:
+
+* Plus Code Inputs MUST treat non-upper-case inputs as if they were upper case.
+* Plus Code outputs MUST be uppercase and conform to the [Plus Codes Specification].
```

Comment:
Comment: "MUST be uppercase" is more strict than "should return upper case characters" in current API.txt. (Similarly for inputs that MUST vs. should be accepted in non-major-case form). I agree that this new requirement is sensible, but it is an incompatible API change in terms of semver.

Reply:
It is currently specified that inputs are case insensitive at https://github.com/google/open-location-code/blob/a323e80159d6d6170158cc22497250cb86bf3583/docs/specification.md#library-implementation-requirements. Not a breaking change

Currently the spec references the digits for [Plus Codes] as upper case at https://github.com/google/open-location-code/blob/a323e80159d6d6170158cc22497250cb86bf3583/docs/specification.md#digits and the API specs say that a [Plus Code] is returned. One interpretation of these words is that outputs MUST be uppercase.

Good notes here, it helped me find and improve an inconsistency with the PCREs at 64ffd9c

Reply:
However, API.txt states that (emphasis mine):

> Note that any method that returns an Open Location Code **should** return
> upper case characters.
> 
> Methods that accept Open Location Codes as parameters **should** be case
> insensitive.

If different parts of the documentation disagree on something, that's bad. However, while that is the case I believe we can't really state that strictly mandating something that was optional at least to some readings of the previous documentation is **not** a breaking change.

Reply:
I can get behind that argument. Fixed at 2e8ac0abc41e71b30747a62970e7c177193ea41f